LEGO1/legogamestate.cpp

```diff
@@ -1,39 +1,261 @@
 }
 
-// OFFSET: LEGO1 0x10039720
+// OFFSET: LEGO1 0x10039720 STUB
 LegoGameState::~LegoGameState()
 {
   // TODO
 }
 
-// OFFSET: LEGO1 0x10039c60
+// OFFSET: LEGO1 0x10039c60 STUB
 MxResult LegoGameState::Load(MxULong)
 {
   // TODO
   return 0;
 }
 
-// OFFSET: LEGO1 0x10039980
-MxResult LegoGameState::Save(MxULong p)
+// OFFSET: LEGO1 0x100f3e40
+const char *g_fileExtensionGS = ".GS";
+
+// OFFSET: LEGO1 0x1003a170
+void LegoGameState::GetFileSavePath(MxString *p_outPath, MxULong p_slotn)
 {
-  // TODO
-  return 0;
+  char baseForSlot[2] = "0";
+  char path[1024] = "";
+
+  // Save path base
+  if (m_savePath != NULL)
+    strcpy(path, m_savePath);
+
+  // Slot: "G0", "G1", ...
+  strcat(path, "G");
+  baseForSlot[0] += p_slotn;
+  strcat(path, baseForSlot);
+
+  // Extension: ".GS"
+  strcat(path, g_fileExtensionGS);
+  *p_outPath = &MxString(path);
+}
+
+struct ColorStringStruct
+{
+  const char *m_targetName;
+  const char *m_colorName;
+};
+
+// OFFSET: LEGO1 0x100f3e58
+ColorStringStruct g_colorSaveData[43] = {
+  {"c_dbbkfny0", "lego red"},
+  {"c_dbbkxly0", "lego white"},
+  {"c_chbasey0", "lego black"},
+  {"c_chbacky0", "lego black"},
+  {"c_chdishy0", "lego white"},
+  {"c_chhorny0", "lego black"},
+  {"c_chljety1", "lego black"},
+  {"c_chrjety1", "lego black"},
+  {"c_chmidly0", "lego black"},
+  {"c_chmotry0", "lego blue"},
+  {"c_chsidly0", "lego black"},
+  {"c_chsidry0", "lego black"},
+  {"c_chstuty0", "lego black"},
+  {"c_chtaily0", "lego black"},
+  {"c_chwindy1", "lego black"},
+  {"c_dbfbrdy0", "lego red"},
+  {"c_dbflagy0", "lego yellow"},
+  {"c_dbfrfny4", "lego red"},
+  {"c_dbfrxly0", "lego white"},
+  {"c_dbhndly0", "lego white"},
+  {"c_dbltbry0", "lego white"},
+  {"c_jsdashy0", "lego white"},
+  {"c_jsexhy0",  "lego black"},
+  {"c_jsfrnty5", "lego black"},
+  {"c_jshndly0", "lego red"},
+  {"c_jslsidy0", "lego black"},
+  {"c_jsrsidy0", "lego black"},
+  {"c_jsskiby0", "lego red"},
+  {"c_jswnshy5", "lego white"},
+  {"c_rcbacky6", "lego green"},
+  {"c_rcedgey0", "lego green"},
+  {"c_rcfrmey0", "lego red"},
+  {"c_rcfrnty6", "lego green"},
+  {"c_rcmotry0", "lego white"},
+  {"c_rcsidey0", "lego green"},
+  {"c_rcstery0", "lego white"},
+  {"c_rcstrpy0", "lego yellow"},
+  {"c_rctailya", "lego white"},
+  {"c_rcwhl1y0", "lego white"},
+  {"c_rcwhl2y0", "lego white"},
+  {"c_jsbasey0", "lego white"},
+  {"c_chblady0", "lego black"},
+  {"c_chseaty0", "lego white"},
+};
+
+// NOTE: This offset = the end of the variables table, the last entry
+// in that table is a special entry, the string "END_OF_VARIABLES"
+// OFFSET: LEGO1 0x100f3e50
+extern const char *s_endOfVariables;
+
+// OFFSET: LEGO1 0x1003a020
+MxResult __stdcall WriteEndOfVariables(LegoStream *p_stream)
+{
+  unsigned char len = strlen(s_endOfVariables);
+  if (p_stream->Write(&len, 1) == SUCCESS)
+    return p_stream->Write(s_endOfVariables, len);
+  return FAILURE;
+}
+
+struct LegoSaveDataEntry3
+{
+  char *m_name;
+  void *m_pSomething1;
```

Comment:
Same here, unknown variables should have their offset, not be numbered sequentially.